File: java/lesson 11/Subsets.java
// https://leetcode.com/problems/subsets/

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        subsets(nums, 0, new LinkedList<Integer>(), result);
        return result;
    }
    
    private void subsets(int[] nums, int start, LinkedList<Integer> current, List<List<Integer>> result) {
        result.add((LinkedList<Integer>) current.clone());
        for (int i = start; i < nums.length; i++)  {
            current.add(nums[i]);
            subsets(nums, i + 1, current, result);
            current.removeLast();
        }
    }
}

File: kotlin/lesson 11/Subsets.kt
// https://leetcode.com/problems/subsets/

class Solution {
    fun subsets(nums: IntArray): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        subsets(nums, 0, LinkedList<Int>(), result)
        return result
    }
    
    fun subsets(nums: IntArray, start: Int, current: LinkedList<Int>, result: MutableList<List<Int>>) {
        result.add(ArrayList(current))
        
        for (i in start..nums.size - 1) {
            current.add(nums[i])
            subsets(nums, i + 1, current, result)
            current.removeLast()
        }
    }
}
